# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0 Late days

In [134]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, classification_report


## 1. Loading the dataset

In [31]:
# Using pandas load the dataset
data = pd.read_csv("science_data_large.csv")
# Output the first 15 rows of the data
print(data.head(15))
print('\n')
# Display a summary of the table information (number of datapoints, etc.)
data.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000

## 2. Splitting the dataset

In [24]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[["Temperature °C", "Mols KCL"]]
Y = data["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.1, random_state=42)


## 3. Perform a Linear Regression

In [47]:
# Use sklearn to train a model on the training set
model = LinearRegression()
reg = model.fit(X_train, Y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
print(reg.predict(pd.DataFrame({'Temperature °C': [100], 'Mols KCL' : [100]})))
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print(reg.score(X_test, Y_test))
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print(reg.coef_)
print(reg.intercept_)

[-219507.33159764]
0.8552472077276095
[ 866.14641337 1032.69506649]
-409391.4795834075


Write the linear equation of a slime: (example equation: $E = mc^2$)

$s$ = Size nm^3

$t$ = Temperature °C

$m$ = Mols KCL

$s = -409391.47958 + 866.14641t + 1032.69507m$ 

Report on score and explain meaning:  
Score: 0.85525

 The score represents the coefficient of determination, which represents how well the model fits the data, and thus how well the features predict the label. We use the test data to get this value. A score of 1.0 means that all of the datapoints lie on the line, and that the features perfectly predict the labels.  A score of 0.0 means that the features do not explain the labels at all, and that just using the mean of the labels would work better. Since our score is 0.85525, this means that 85.525% of the variance in the size of the slime can be predicted by the temperature and amount of potassium chloride used on the slime.

## 4. Use Cross Validation

In [55]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kf = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(model, X, Y, cv=kf)

# Report on their finding and their significance
print(scores)
scores.mean()

[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]


0.8597294202684646

Write findings here:
My array of scores was 
0.86151, 0.82742, 0.87195, 0.88166, and 0.85609
Which has an average score of
0.85973
These represent the coefficients of determination for each of the 5 folds of the data. The data is divided into 5 folds, and five separate models are made such that a different fold (20% of data) is used as the test set for the model, while the rest are used to train the model. The coefficients of determinations were found by finding the results of using each fold as test data on their corresponding models.  Since each of the coefficients are between 0.82 and 0.89, we can estimate that the real coefficient of determination falls somewhere in that range, and is likely around 0.85973.

## 5. Using Polynomial Regression

In [75]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
newData = PolynomialFeatures(2)
newX = newData.fit_transform(X)


newScores = []
coefficients = []

for train_index, test_index in kf.split(newX):
    x_train, x_test = newX[train_index], newX[test_index]
    y_train, y_test = Y[train_index], Y[test_index]

    # Create and fit the linear regression model
    model = LinearRegression()
    newReg = model.fit(x_train, y_train)

    # Store the coefficients for each fold
    coefficients.append(newReg.coef_)  
    newScores.append(newReg.score(x_test, y_test))
# Convert coefficients list to a DataFrame for easier analysis
coeffs_df = pd.DataFrame(coefficients, columns = newData.get_feature_names_out())
newScores_df = pd.DataFrame(newScores)

# Report on the metrics and output the resultant equation as you did in Part 3.

print(coefficients)

print(newScores)


[array([ 0.00000000e+00,  1.20000000e+01, -1.32006599e-07, -1.59802454e-11,
        2.00000000e+00,  2.85714287e-02]), array([ 0.00000000e+00,  1.20000000e+01, -1.39793270e-07,  1.66127426e-11,
        2.00000000e+00,  2.85714287e-02]), array([ 0.00000000e+00,  1.20000000e+01, -1.04364350e-07,  1.98306044e-11,
        2.00000000e+00,  2.85714287e-02]), array([ 0.00000000e+00,  1.20000000e+01, -1.09881154e-07, -4.56258730e-11,
        2.00000000e+00,  2.85714287e-02]), array([ 0.00000000e+00,  1.20000000e+01, -1.33837141e-07, -2.71315121e-11,
        2.00000000e+00,  2.85714287e-02])]
[1.0, 1.0, 1.0, 1.0, 1.0]


Write the polynomial equation of a slime: (example equation: $E = mc^2$) 

Using the average coefficients from each test (Since the coefficients remained the same in every model when rounded to the 5th decimal place)

$s$ = Size nm^3

$t$ = Temperature °C

$m$ = Mols KCL

$s = 12t + 2tm + 0.02857m^2$

Report on the score and interpret:


The average score from cross validation was 1.0. this means that this new model that takes into account polynomial values of temperature and amount of potassium chloride used, perfectly predicts what the size of the slime is going to be.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [138]:
# Load the dataset. Then train and evaluate the classification models.
data = pd.read_csv("ckd_feature_subset.csv")

X = data[["age", "bp", 'wbcc', 'rbcc', 'appet_poor', 'appet_good']]
Y = data[['Target_ckd']].values
Y = Y.ravel()

kf = KFold(n_splits=5, shuffle=True, random_state=42)

logModel = LogisticRegression(max_iter=200)
log_reg_scores = cross_val_score(logModel, X, Y, cv=kf) 
log_reg_score = log_reg_scores.mean()
log_reg_SD = log_reg_scores.std()
print("Logistic Regression Score Mean : ", log_reg_score, " Standard Deviation : ", log_reg_SD)

svcModel = SVC()
svc_scores = cross_val_score(svcModel, X, Y, cv=kf) 
svc_score = svc_scores.mean()
svc_SD = svc_scores.std()
print("Support Vector Machines Score Mean : ", svc_score, " Standard Deviation : ", svc_SD)


neighborsModel = KNeighborsClassifier(n_neighbors = 3)
neighbors_scores = cross_val_score(neighborsModel, X, Y, cv=kf) 
neighbors_score = neighbors_scores.mean()
neighbors_SD = neighbors_scores.std()
print("k-Nearest Neighbors Score Mean : ", neighbors_score, " Standard Deviation : ", neighbors_SD)

NNModel = MLPClassifier(hidden_layer_sizes=(100, ), max_iter= 200, random_state=42)
NN_scores = cross_val_score(NNModel, X, Y, cv=kf) 
NN_score = NN_scores.mean()
NN_SD = NN_scores.std()
print("Neural Network Score Mean : ", NN_score, " Standard Deviation : ", NN_SD)

Logistic Regression Score Mean :  0.8565591397849461  Standard Deviation :  0.0662686697755876
Support Vector Machines Score Mean :  0.9281720430107526  Standard Deviation :  0.04760126519964333
k-Nearest Neighbors Score Mean :  0.9475268817204301  Standard Deviation :  0.03936952543394417


/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warning

Neural Network Score Mean :  0.8698924731182794  Standard Deviation :  0.07329516870241994


/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments
| Model | Average Score | Standard Deviation of Scores |
|----------|----------|----------|
| Logistical Regression | 0.85656 | 0.06627 |
| Support Vector Machines | 0.92817 | 0.04760 |
| k-Nearest Neighbors | 0.94753 | 0.03937 |
| Neural Network | 0.86989 | 0.07330 |

Every model had a high coefficient of determination, and a low standard deviation (under 0.1). Their coefficients of determination were all quite close, but the one that performed best was the k-Nearest neighbors.

The model that performed best was the k-nearest neighbors model. This makes sense since k-nn usually performs best on datasets with lower dimensions, and with smaller datasets. It can handle non-linear data well, unlike logistic regression. In addition, there were a lot of clusters, especially in bp, and the binary appet variables, which k-nn works well with. That being said, I think the neural network may eventually be able to perform better if I tune it more.

In [146]:
# Experiments with Neural Network.

NNModel1 = MLPClassifier(hidden_layer_sizes=(100, ), max_iter= 5000, random_state=42)
NN_scores1 = cross_val_score(NNModel1, X, Y, cv=kf) 
NN_score1 = NN_scores1.mean()
NN_SD1 = NN_scores1.std()
print("Neural Network 1 Score Mean : ", NN_score1, " Standard Deviation : ", NN_SD1)


NNModel2 = MLPClassifier(hidden_layer_sizes=(1000, ), max_iter= 200, random_state=42)
NN_scores2 = cross_val_score(NNModel2, X, Y, cv=kf) 
NN_score2 = NN_scores2.mean()
NN_SD2 = NN_scores2.std()
print("Neural Network 2 Score Mean : ", NN_score2, " Standard Deviation : ", NN_SD2)

NNModel3 = MLPClassifier(hidden_layer_sizes=(10, 10, 10, 10, 10, 10, 10, 10, 10, 10), max_iter= 200, random_state=42)
NN_scores3 = cross_val_score(NNModel3, X, Y, cv=kf) 
NN_score3 = NN_scores3.mean()
NN_SD3 = NN_scores3.std()
print("Neural Network Score Mean 3 : ", NN_score3, " Standard Deviation : ", NN_SD3)

NNModel4 = MLPClassifier(hidden_layer_sizes=(10, 10, 10, 10, 10, 10, 10, 10, 10, 10), max_iter= 5000, random_state=42)
NN_scores4 = cross_val_score(NNModel4, X, Y, cv=kf) 
NN_score4 = NN_scores4.mean()
NN_SD4 = NN_scores4.std()
print("Neural Network Score Mean 4 : ", NN_score4, " Standard Deviation : ", NN_SD4)

Neural Network 1 Score Mean :  0.9415053763440862  Standard Deviation :  0.04717771803967577


/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warning

Neural Network 2 Score Mean :  0.9350537634408603  Standard Deviation :  0.040465552982107675


/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warning

Neural Network Score Mean 3 :  0.9739784946236559  Standard Deviation :  0.024200954660877776
Neural Network Score Mean 4 :  0.9739784946236559  Standard Deviation :  0.024200954660877776


## Results and Conclusion for Neural Network Experiments

| Neural Network | Average Score | Standard Deviation of Scores |
|----------|----------|----------|
| 1 | 0.94151 | 0.04718 |
| 2 | 0.93505 | 0.04047 |
| 3 | 0.97398 | 0.02420 |
| 4 | 0.97398 | 0.02420 |


I played with a few different variables. The first was the amount of iterations. I increased it by 10 times, and found a substantial bump in the score. 

I then did a second test, but increased the amount of nodes by 10 times, and found a smaller but still big bump in the score. 

Next, I kept the overall amount of nodes the same, but made 10 layers of 10 nodes instead. This caused the greatest impact on the score. I then increased the number of iterations on this 10 layer network, but found that the score remained the same. Overall, I found the measure that had the greatest impact on the effectiveness of the algorithm was the number of layers.